File: tools/toolchain/proto_contract_test.go
package toolchain_test

import (
	"testing"

	storagev1 "github.com/sagar-arora/veclet/gen/go/veclet/storage/v1"
	vecletv1 "github.com/sagar-arora/veclet/gen/go/veclet/v1"
	"google.golang.org/protobuf/proto"
)

func TestGeneratedContractsRoundTripCanonicalStoredRecord(t *testing.T) {
	stored := &storagev1.StoredRecord{
		LocalIndexId: 7,
		Record: &vecletv1.VectorRecord{
			VectorId:    "customer-42",
			Version:     1,
			Embedding:   []float32{0.25, -0.5},
			PayloadData: "opaque payload",
		},
	}

	encoded, err := proto.Marshal(stored)
	if err != nil {
		t.Fatalf("marshal StoredRecord: %v", err)
	}

	decoded := new(storagev1.StoredRecord)
	if err := proto.Unmarshal(encoded, decoded); err != nil {
		t.Fatalf("unmarshal StoredRecord: %v", err)
	}
	if got, want := decoded.GetRecord().GetVectorId(), "customer-42"; got != want {
		t.Fatalf("vector ID = %q, want %q", got, want)
	}
	if got, want := len(decoded.GetRecord().GetEmbedding()), 2; got != want {
		t.Fatalf("embedding length = %d, want %d", got, want)
	}
}

func TestGeneratedContractsExposeExpectedServices(t *testing.T) {
	if got, want := vecletv1.QueryService_ServiceDesc.ServiceName, "veclet.v1.QueryService"; got != want {
		t.Fatalf("query service = %q, want %q", got, want)
	}
	if got, want := vecletv1.DataService_ServiceDesc.ServiceName, "veclet.v1.DataService"; got != want {
		t.Fatalf("data service = %q, want %q", got, want)
	}
}

File: engine/tests/proto_contract_test.cc
#include <string>

#include <gtest/gtest.h>

#include "veclet/storage/v1/record.pb.h"
#include "veclet/v1/data.grpc.pb.h"
#include "veclet/v1/query.grpc.pb.h"

namespace {

TEST(GeneratedProtoContract, RoundTripsCanonicalStoredRecord) {
  veclet::storage::v1::StoredRecord stored;
  stored.set_local_index_id(7);
  stored.mutable_record()->set_vector_id("customer-42");
  stored.mutable_record()->set_version(1);
  stored.mutable_record()->add_embedding(0.25F);
  stored.mutable_record()->add_embedding(-0.5F);
  stored.mutable_record()->set_payload_data("opaque payload");

  std::string encoded;
  ASSERT_TRUE(stored.SerializeToString(&encoded));

  veclet::storage::v1::StoredRecord decoded;
  ASSERT_TRUE(decoded.ParseFromString(encoded));
  EXPECT_EQ(decoded.local_index_id(), 7);
  EXPECT_EQ(decoded.record().vector_id(), "customer-42");
  EXPECT_EQ(decoded.record().embedding_size(), 2);
}

TEST(GeneratedProtoContract, ExposesExpectedServices) {
  EXPECT_EQ(veclet::v1::QueryService::service_full_name(),
            "veclet.v1.QueryService");
  EXPECT_EQ(veclet::v1::DataService::service_full_name(),
            "veclet.v1.DataService");
}

}  // namespace
